ingest/lambdas/rss-feed-ingester.js

```diff
@@ -0,0 +1,92 @@
+/* eslint-disable import/no-unresolved, import/extensions */
+// @ts-check
+const pMap = require('p-map');
+
+const dspaceClient = require('../../lib/dspace-client');
+const ingestQueue = require('../lib/ingest-queue');
+const { getStringFromEnv } = require('../../lib/env-utils');
+
+const dspaceFeedReadInterval = process.env.DSPACE_FEED_READ_INTERVAL
+  ? Number.parseInt(process.env.DSPACE_FEED_READ_INTERVAL)
+  : 300;
+
+/**
+ * Returns a boolean indiciting whether or not we shoudl publish documents
+ * from the RSS feed based on when the RSS feed was published and the
+ * last time we checked it. Set the `feedReadInterval` to -1 to always
+ * pass this test.
+ *
+ * @param {Date} publishedDate The published date of the RSS feed.
+ * @param {number} feedReadInterval The number in seconds since the
+ * published RSS feed was last read.
+ *
+ * @returns {boolean} Whether or not we should publish new documents based
+ *          on the last time we checked the RSS feed
+ */
+const shouldQueuePublishedDocuments = (publishedDate, feedReadInterval) => {
+  if (feedReadInterval === -1) {
+    return true;
+  }
+
+  const feedReadIntervalDate = new Date();
+  feedReadIntervalDate.setSeconds(feedReadInterval * -1);
+
+  return publishedDate >= feedReadIntervalDate;
+};
+
+/**
+ * Fetches a DSpace RSS feed and determines whether or not the documents
+ * listed in the feed should be queued for ingest. This is determined by
+ * checking the published date of the feed, and whether or not it has been
+ * published (updated) since the last time we checked it.
+ */
+const handler = async () => {
+  const dspaceEndpoint = getStringFromEnv('DSPACE_ENDPOINT');
+  const ingestTopicArn = getStringFromEnv('INGEST_TOPIC_ARN');
+
+  try {
+    // Fetch the RSS feed from DSpace.
+    const feed = await dspaceClient.getFeed(dspaceEndpoint);
+
+    // Parse out the last published date and use it to determine whether or
+    // not we should index new documents.
+    const lastPublishedDate = new Date(feed.channel[0].pubDate[0]._);
+    if (!shouldQueuePublishedDocuments(
+      lastPublishedDate,
+      dspaceFeedReadInterval
+    )) {
+      return;
+    }
+
+    // Loop through the feed items and queue them for ingest. In order to do
+    // this we have to fetch the metadata for each item based on the
+    // dc.identifier.uri field since that's all that is available in the feed.
+    // This identifier doesn't have a consistent domain so we use the UUID for
+    // ingest queuing.
+    await pMap(
```

Comment:
Consider creating `lib/sns-utils.ts` with:

```ts
import { chunk } from 'lodash';
import pMap from 'p-map';
import { sns } from './aws-clients';

interface PublishBatchInput {
  topicArn: string
  messages: string[],
  concurrency?: number
}

export const publishBatch = async (
  params: PublishBatchInput
): Promise<void> => {
  const { concurrency, messages, topicArn: TopicArn } = params;

  await pMap(
    chunk(messages, 10),
    (messageChunk) => {
      const PublishBatchRequestEntries = messageChunk.map((Message, index) => ({
        Id: index.toString(),
        Message,
      }));

      return sns().publishBatch({
        TopicArn,
        PublishBatchRequestEntries,
      }).promise();
    },
    { concurrency }
  );
};
```

With that utility, this could be refactored to something like:

```ts
  const dSpaceItems = await pMap(
      feed.channel[0].item,
      (feedItem) => dspaceClient.find(
        dspaceEndpoint,
        'dc.identifier.uri',
        feedItem.link[0]
      ),
      { concurrency: 5 }
    );

    const uuids = dSpaceItems.flat().map((i) => i.uuid);

    await publishBatch({
      topicArn: ingestTopicArn,
      messages: uuids,
      concurrency: 5,
    });
```

That feels a bit easier to read.

Reply:
@marchuffnagle Could we file this as a Github issue? I think this is a good idea and I'd like to get it working with Localstack, etc. but we'd have to make sure the tests are updated and and configure the SNS client in `aws-clients` – all good things but considering the RSS reader is fairly trivial code (and we're only ever dealing with at most 5 items in the feed anyway) maybe it's better to circle back around to this later?

Reply:
> Could we file this as a GitHub issue?

Sure.